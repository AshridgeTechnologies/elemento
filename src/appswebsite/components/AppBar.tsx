import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}