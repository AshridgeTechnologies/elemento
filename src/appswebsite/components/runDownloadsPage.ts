import thePage from './DownloadsPage'
import {run} from '../../appsShared/renderInPage'

run(thePage)
