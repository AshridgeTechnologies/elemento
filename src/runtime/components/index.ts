export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
