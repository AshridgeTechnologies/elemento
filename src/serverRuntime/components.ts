export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts