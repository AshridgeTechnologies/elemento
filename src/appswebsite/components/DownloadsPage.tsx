import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
