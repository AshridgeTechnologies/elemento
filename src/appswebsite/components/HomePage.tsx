import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import HomeIcon from '@mui/icons-material/Home'
import LightbulbIcon from '@mui/icons-material/LightbulbTwoTone'
import CreateIcon from '@mui/icons-material/CreateTwoTone'
import AodIcon from '@mui/icons-material/AodTwoTone'
import ArrowCircleRightTwoToneIcon from '@mui/icons-material/ArrowCircleRightTwoTone';

const primary = theme.palette.primary.main
const secondary = theme.palette.secondary.main

function Logo() {
    return <Box color={secondary} border='8px solid' borderRadius={4} width={108} height={104} paddingTop='10px' paddingX='3px'>
        <Stack>
            <Typography fontSize='100px' fontFamily='"Helvetica"' align='center' lineHeight={0.8}>El</Typography>
            <Typography fontSize='16px' lineHeight={1}>42</Typography>
        </Stack>
    </Box>
}

function SectionLink(props: {title: string, href: string, icon: any, children: any}) {
    const Icon = props.icon
    return <Box border='4px solid' borderColor={primary} borderRadius={4} width='300px' paddingY='10px' paddingX='16px'>
        <Link href={props.href} underline='none'>
            <Stack spacing={2}>
                <Box textAlign='center'><Icon sx={{ fontSize: 180 }} /></Box>
                <Link component='div' underline='hover' href={props.href} variant='h5' textAlign='center'>{props.title}</Link>
                <Typography fontSize='20px' color='#555'>{props.children}</Typography>
                <Box textAlign='center'><ArrowCircleRightTwoToneIcon sx={{ fontSize: 32, color: secondary }} /></Box>
            </Stack>
        </Link>
    </Box>
}

export default function HomePage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Home'/>
            </Box>
            <Box flex='1' minHeight={0} mt={2}>
                <Container maxWidth="lg">
                    <Stack spacing={2}>
                        <Stack direction='row'>
                            <Logo/>
                            <Stack ml={4}>
                                <Typography fontWeight='bold' variant='h3'  color={secondary} align='center'>Elemento</Typography>
                                <Typography variant='h4' color={secondary} align='center' mt={2}>
                                    Everything you need to make and use your own software
                                </Typography>
                            </Stack>
                        </Stack>
                        <Stack direction='row' justifyContent='space-between' alignItems='stretch' marginTop='30px'>
                            <SectionLink title='Learn' href='/learn' icon={LightbulbIcon}>
                                A great range of resources to teach you how to get started and create your first app with Elemento.
                            </SectionLink>
                            <SectionLink title='Create' href='/studio' icon={CreateIcon}>
                                Put your app together and preview it live in Elemento Studio
                            </SectionLink>
                            <SectionLink title='Use' href='/run' icon={AodIcon}>
                                Run your app - or go back to the apps you have already made.
                            </SectionLink>
                        </Stack>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
