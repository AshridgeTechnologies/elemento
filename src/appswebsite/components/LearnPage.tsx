import React from 'react'
import Toolbar from '@mui/material/Toolbar'
import {Box, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import {TutorialRunner} from '../../tutorials/TutorialRunner'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Learn' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <TutorialRunner/>
            </Box>
        </Box>
    </ThemeProvider>
}
