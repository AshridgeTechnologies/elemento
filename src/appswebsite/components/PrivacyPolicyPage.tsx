import React from 'react'
import {Box, Stack, Typography} from '@mui/material'
import AppBar from './AppBar'

export default function PrivacyPolicyPage() {

    return <Box display='flex' flexDirection='column' height='100%' width='100%'>
        <Box flex='0'>
            <AppBar/>
        </Box>
        <Box flex='1' minHeight={0}>
            <Stack spacing={2}>
                <Typography variant='h4'>Elemento Apps - Privacy Policy</Typography>
                <Typography variant='body1'>Coming soon</Typography>
            </Stack>
        </Box>
    </Box>
}
