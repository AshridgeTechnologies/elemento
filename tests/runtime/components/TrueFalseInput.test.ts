/**
 * @vitest-environment jsdom
 */

import {expect, test} from 'vitest'
import {TrueFalseInput} from '../../../src/runtime/components/index'
import {DEBUG_TIMEOUT, snapshot, testAppInterface, valueObj, wait, wrappedTestElement} from '../../testutil/testHelpers'
import {testContainer} from '../../testutil/rtlHelpers'

const [trueFalseInput, appStoreHook] = wrappedTestElement(TrueFalseInput)

const stateAt = (path: string) => appStoreHook.stateAt(path)

test('TrueFalseInput element produces output with properties supplied',
    snapshot(trueFalseInput('app.page1.width', {initialValue: true, label: 'Covered', styles: {color: 'red'}}))
)

test('TrueFalseInput element produces output with default values where properties omitted',
    snapshot(trueFalseInput('app.page1.height', {}))
)

test('TrueFalseInput shows false if that value is in the state', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget1', {initialValue: false}))
    expect(el`app.page1.widget1`.checked).toBe(false)
})

test('TrueFalseInput shows true if that value is in the state', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget1', {initialValue: true}), 'container2')
    expect(el`app.page1.widget1`.checked).toBe(true)
})

test('TrueFalseInput shows false if no value is in the state', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget1x', {}))
    expect(el`app.page1.widget1x`.checked).toBe(false)
})

test('TrueFalseInput shows false if an undefined value is in the state', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget11', {value: undefined}), 'container4')
    expect(el`app.page1.widget11`.checked).toBe(false)
})

test('TrueFalseInput shows initial value when state value is set to undefined and initial value exists', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget1', {initialValue: true}), 'container3')
    expect(el`app.page1.widget1`.checked).toBe(true)
})

// TODO check reason for this test
test.skip('TrueFalseInput shows empty value when state value is set to null and initial value exists', async () => {
    const {el, user} = testContainer(trueFalseInput('app.page1.widget1', {initialValue: true}), 'testContainer2')
    expect(el`app.page1.widget1`.checked).toBe(true)
    stateAt('app.page1.widget1')._setValue(null)
    await wait()
    expect(el`app.page1.widget1`.checked).toBe(false)

})

test('TrueFalseInput shows true value when state value is set to true and initial value exists', async () => {
    const {el, user} = testContainer(trueFalseInput('app.page1.widget22', {initialValue: false}), 'testContainer3')
    expect(el`app.page1.widget22`.checked).toBe(false)
    stateAt('app.page1.widget22')._setValue(true)
    await wait(20)
    expect(el`app.page1.widget22`.checked).toBe(true)
})

test('TrueFalseInput shows false value when state value is set to false and initial value exists', async () => {
    const {el, user} = testContainer(trueFalseInput('app.page1.widget77', {initialValue: true}), 'testContainer4')
    expect(el`app.page1.widget77`.checked).toBe(true)
    stateAt('app.page1.widget77')._setValue(false)
    await wait(20)
    expect(el`app.page1.widget77`.checked).toBe(false)

})

test('TrueFalseInput element produces output with other properties supplied as state objects', () => {
    const {el} = testContainer(trueFalseInput('app.page1.widget1', {initialValue: true, label: valueObj('Is it OK')}), 'container5')
    expect(el`label`.textContent).toBe('Is it OK')
})

test('TrueFalseInput stores updated values in the app store section for its path', async () => {
    const {el, user} = testContainer(trueFalseInput('app.page1.sprocket', {initialValue: false}))
    const inputEl = el`app.page1.sprocket`
    await user.click(inputEl)
    expect(stateAt('app.page1.sprocket').value).toBe(true)
} )

test('TrueFalseInput does not change if readonly', async () => {
    const {el, user} = testContainer(trueFalseInput('app.page1.sprocket', {initialValue: false, readOnly: true}))
    const inputEl = el`app.page1.sprocket`
    await expect(user.click(inputEl)).rejects.toThrow(/Unable to perform pointer interaction/)
    expect(stateAt('app.page1.sprocket').value).toBe(false)
} )

test('State class has correct properties', () => {
    const emptyState = new TrueFalseInput.State({})
    expect(emptyState.value).toBe(false)
    expect(emptyState.toString()).toBe('false')
    expect(emptyState.dataValue).toBe(null)
    expect(emptyState.defaultValue).toBe(false)

    const state = new TrueFalseInput.State({initialValue: true})
    const appInterface = testAppInterface('testPath', state)
    expect(state.value).toBe(true)
    expect(state.toString()).toBe('true')

    state.Reset()
    const resetState = state._withStateForTest({value: undefined, errorsShown: false})
    expect(appInterface.updateVersion).toHaveBeenCalledWith(state.withMergedState({value: undefined, errorsShown: false}))
    expect(resetState.value).toBe(true)
    expect(resetState.dataValue).toBe(true)
})

test('State is modified only when its value is not null and different to its empty initial value', () => {
    const state = new TrueFalseInput.State({})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: true})
    expect(updatedState.modified).toBe(true)
    const resetState = updatedState._withStateForTest({value: undefined})
    expect(resetState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(false)
    const updatedToFalseState = updatedState._withStateForTest({value: false})
    expect(updatedToFalseState.modified).toBe(false)
})

test('State is modified only when its value is not null and different to its true initial value', () => {
    const state = new TrueFalseInput.State({initialValue: true})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: false})
    expect(updatedState.modified).toBe(true)
    const resetState = updatedState._withStateForTest({value: undefined})
    expect(resetState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(true)
})

test('State is modified only when its value is different to its false value', () => {
    const state = new TrueFalseInput.State({initialValue: false})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: true})
    expect(updatedState.modified).toBe(true)
    const resetState = updatedState._withStateForTest({value: undefined})
    expect(resetState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(false)
})
