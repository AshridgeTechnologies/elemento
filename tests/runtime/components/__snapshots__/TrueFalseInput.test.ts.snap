// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrueFalseInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
>
  <label
    className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart css-1intw90-MuiFormControlLabel-root"
    htmlFor="app.page1.height"
  >
    <span
      className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall css-vuhw6e-MuiButtonBase-root-MuiSwitchBase-root-MuiCheckbox-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        checked={false}
        className="PrivateSwitchBase-input css-12xagqm-MuiSwitchBase-root"
        data-indeterminate={false}
        disabled={false}
        id="app.page1.height"
        onChange={[Function]}
        readOnly={false}
        required={false}
        type="checkbox"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
    <span
      className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
      style={{}}
    />
  </label>
  <p
    className="MuiFormHelperText-root MuiFormHelperText-sizeMedium css-18y2hez-MuiFormHelperText-root"
  />
</div>
`;

exports[`TrueFalseInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
>
  <label
    className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart css-1intw90-MuiFormControlLabel-root"
    htmlFor="app.page1.width"
  >
    <span
      className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall css-135xnp2-MuiButtonBase-root-MuiSwitchBase-root-MuiCheckbox-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        checked={true}
        className="PrivateSwitchBase-input css-12xagqm-MuiSwitchBase-root"
        data-indeterminate={false}
        disabled={false}
        id="app.page1.width"
        onChange={[Function]}
        readOnly={false}
        required={false}
        type="checkbox"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
        data-testid="CheckBoxIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        />
      </svg>
    </span>
    <span
      className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
      style={{}}
    >
      Covered
    </span>
  </label>
  <p
    className="MuiFormHelperText-root MuiFormHelperText-sizeMedium css-18y2hez-MuiFormHelperText-root"
  />
</div>
`;
